### PR TITLE
Add extension for document-scoped search

### LINKED ISSUE #181: Find and Replace
How to find and replace texts globally? I can only do within-cell search by press "cmd + F".

### PR DESCRIPTION
**Pull request by loichuder**
Fix #181 

Well, that wasn't as easy as I first thought :sweat_smile: 

I could not reuse `@jupyterlab/documentsearch-extension` plugin directly as it relies on `ILabShell` to find the current searchable widget. I therefore rewrote the plugin for retrolab in a new `@retrolab/documentsearch-extension` to account for retrolab's differences.

I still have a small issue on RetroLab's landing page where `Ctrl+F` is caught when the focus is on the `Files` tab when it should not. This is perhaps related to the way I handled `currentWidget` in the new plugin.

**EDIT: Now fixed !**

Anyway, before going further, I wanted to draft this first functional implementation to know if I was going the right way. Any comments/feedback are welcomed.

### CODE REVIEW

File: packages/documentsearch-extension/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+const retroShellWidgetListener: JupyterFrontEndPlugin<void> = {
+  id: '@retrolab/documentsearch-extension:retroShellWidgetListener',
+  requires: [IRetroShell, ISearchProviderRegistry],
+  autoStart: true,
+  activate: (
+    app: JupyterFrontEnd,
+    retroShell: IRetroShell,
+    registry: ISearchProviderRegistry
+  ) => {
```
Comment (loichuder) [inline packages/documentsearch-extension/src/index.ts:20]: Reimplementation of https://github.com/jupyterlab/jupyterlab/blob/a8dff39a1a2be850a80861e05145e763079b8794/packages/documentsearch-extension/src/index.ts#L38 for retrolab

File: packages/documentsearch-extension/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    // changes, in case a provider for the current widget was added
+    // or removed
+    registry.changed.connect(() =>
+      transformWidgetSearchability(retroShell.currentWidget)
+    );
+
+    // Apply the searchable class only to the active widget if it is actually
+    // searchable. Remove the searchable class from a widget when it's
+    // no longer active.
+    retroShell.currentChanged.connect((_, args) => {
```
Comment (loichuder) [inline packages/documentsearch-extension/src/index.ts:46]: Contrary to `labShell.activeChanged`, `retroShell.currentChanged` does not provide `args` containing the old and new widgets. I could not reproduce the behaviour in https://github.com/jupyterlab/jupyterlab/blob/a8dff39a1a2be850a80861e05145e763079b8794/packages/documentsearch-extension/src/index.ts#L64 where the class is removed for inactive widgets.

But, given that Retrolab only has one widget by browser tab, I think this is fine ? :shrug:

**Comment by jtpio** [comment]
Thanks @loichuder for starting this!

> I could not reuse @jupyterlab/documentsearch-extension plugin directly as it relies on ILabShell to find the current searchable widget. I therefore rewrote the plugin for retrolab in a new @retrolab/documentsearch-extension to account for retrolab's differences.

Right, this is something we often hit over the past couple of months when trying to reuse upstream plugins in alternate JupyterLab distributions.

Usually it's fine to duplicate the logic here as temporary solution, and in the meantime we can update the upstream plugin to make is more reusable. In this case this likely means making `ILabShell` optional.

File: app/index.js [hunk 3]
```diff
@@ -123,6 +124,9 @@ async function main() {
       ].includes(id)
     ),
     require('@jupyterlab/docprovider-extension'),
+    require('@jupyterlab/documentsearch-extension').default.filter(({ id }) =>
+      ['@jupyterlab/documentsearch:plugin'].includes(id)
```
Comment (jtpio) [inline app/index.js:128]: It's good that we are at least able to reuse this plugin :+1:

**Comment by jtpio** [comment]
UI Tests failure looks relevant, as the Edit menu now shows new entries:

image

So we can update the reference snapshots with the ones generated in https://github.com/jupyterlab/retrolab/actions/runs/1506929915

**Comment by jtpio** [comment]
> Would it be possible that @jupyterlab/documentsearch-extension:labShellWidgetListener only requires a shell that implements the needed features (activeWidget and activeChanged) 

Maybe, or this could be hoisted to the `JupyterFrontEnd.IShell` base interface although for now we would like to not increase the API surface of `JupyterFrontEnd.IShell`.

**Comment by loichuder** [comment]
@jtpio This is now ready for review.

Could you explain how to update the screenshots ? Perhaps we could also add this information to CONTRIBUTING if that's something maintainers can do ?

**Comment by jtpio** [comment]
> Could you explain how to update the screenshots ? Perhaps we could also add this information to CONTRIBUTING if that's something maintainers can do ?

Yes absolutely. For reference there is some documentation in another repo: https://github.com/jupyterlite/jupyterlite/blob/main/CONTRIBUTING.md#reference-image-captures

But we should have something similar here too. Or as an alternative in the JupyterLab contribution docs and we link to it from here: https://github.com/jupyterlab/jupyterlab/blob/master/docs/source/developer/contributing.rst#visual-regression-and-ui-tests

**Comment by loichuder** [comment]
CI passed :tada:

**Comment by jtpio** [comment]
Thanks!

This looks good on Binder:

https://user-images.githubusercontent.com/591645/145185332-c3d42169-698c-42b6-be0f-7d43a4a62a7a.mp4

We can document updating the reference snapshots in a separate PR.

**Comment by loichuder** [comment]
> We can document updating the reference snapshots in a separate PR.

Yes that was my plan :+1:

**Comment by jtpio** [comment]
cc @yuvipanda if you have any input on this.

I remember you raised some concerns on apps overriding default browser keyboard shortcuts. Here that would be <kbd>Ctrl / Cmd - F</kbd>

If this comes to much in the way, we could also make it an opt-in instead, for example via a `PageConfig` option.

**Comment by yuvipanda** [comment]
Thank you very much for the ping, @jtpio! I really appreciate it! I am also really grateful you working on this, @loichuder!

The browser's default Cmd+F is very helpful, as it matches the mental model of 'search for any text I can see' - including all the *UI elements*. This is very helpful in cases where the user is trying to follow instructions like 'Find item X' - this works even when you just have menu items open (for example):

image

IMO we shouldn't override default browser functionality. The find / replace stuff is pretty cool, but I'd suggest it not be bound to Cmd-F by default. It can be accessed from the menu (as you can in classic notebook right now), and a PageConfig can make that binding if necessary.

**Comment by loichuder** [comment]
> IMO we shouldn't override default browser functionality. The find / replace stuff is pretty cool, but I'd suggest it not be bound to Cmd-F by default.

Make sense to me. 

As I understand, the keyboard shortcut is set via [JupyterLab's document-search extension](https://github.com/jupyterlab/jupyterlab/blob/master/packages/documentsearch-extension/schema/plugin.json#L33) as I reuse one of the plugins of the extension in my implementation.

Is this true ? If so, what would be the way to override this ?

**Comment by jtpio** [comment]
Normally we should be able to disable the shortcut here in retro.

For example when testing this PR on Binder:

https://user-images.githubusercontent.com/591645/145219880-e375c025-bf4f-4bba-806d-2b53e2e535e1.mp4

Which means we could put the following snippet:

```json
{
    "shortcuts": [
        {
            "command": "documentsearch:start",
            "keys": [
                "Accel F"
            ],
            "selector": ".jp-mod-searchable",
            "disabled": true
        }
    ]
}
```

In the retro settings by default.

Reusing the settings override system added in https://github.com/jupyterlab/retrolab/pull/289 so we can customize `jupyter.lab.shortcuts`.

**Comment by jtpio** [comment]
Maybe this could be implemented as a separate plugin with its own settings similar to the `menu` plugin:

https://github.com/jupyterlab/retrolab/blob/49b9ae3f6182dc6311d74b28024d099ca06037da/packages/application-extension/src/index.ts#L199-L229

So it could be disabled with `jupyter labextension disable` by end users.

**Comment by jtpio** [comment]
Also when we implement the override for the `disabled` keyboard shortcut, we could also add a UI test to make sure updating to future versions of `@jupyterlab` packages doesn't break this.

For example this could mean using the `await page.keyboard.press('Control+F)` API as documented here: https://playwright.dev/docs/api/class-keyboard/
And making sure the search / replace popup is not displayed.

It would be fine to add that test in a follow-up PR if we want to merge this PR earlier.

**Comment by loichuder** [comment]
> It would be fine to add that test in a follow-up PR if we want to merge this PR earlier.

If disabling Ctrl+F is the behaviour we want to have, I think it is better to add it to this PR. I just need to make the override work :smile:

**Comment by loichuder** [comment]
Sorry it took me a while to find the courage to navigate once again in the treacherous plugin sea.

I added a plugin to disable the shortcut: 4aed598bf345ced6637a149c67c7f419b0f27478. Not sure if that is the way you intended it @jtpio so don't hesitate to direct me towards another solution. I will work on the UI test, next.

Funfact: While testing, I noticed that the Ctrl+F event is already caught for text files like MD, making a custom search bar appear:
image

This comes **not** from the `documentsearch-extension` but from the notebook classic ! 

So even by disabling the shortcut, the only way to get the browser search for text files is to click out from the document before hitting Ctrl+F.

**Comment by jtpio** [comment]
> I added a plugin to disable the shortcut: [4aed598](https://github.com/jupyterlab/retrolab/commit/4aed598bf345ced6637a149c67c7f419b0f27478). Not sure if that is the way you intended it @jtpio so don't hesitate to direct me towards another solution. I will work on the UI test, next.

Thanks @loichuder. I was more thinking about using the declarative way and override the shortcut in the schema file with `jupyter.lab.shortcuts`. Although I haven't tested yet and the end result is likely to be the same.

> Funfact: While testing, I noticed that the Ctrl+F event is already caught for text files like MD, making a custom search bar appear:

Probably this one would be fine to keep as is yes, especially since classic behaves the same.

**Comment by jtpio** [comment]
FYI @loichuder I pushed the following commit to switch to the declarative approach: https://github.com/jupyterlab/retrolab/pull/294/commits/55915e63e88331197f2b952cdaef69b805697ad6

And rebased to grab the changes from https://github.com/jupyterlab/retrolab/pull/309. Will follow-up with the snapshots update if needed.

**Comment by jtpio** [comment]
> switch to the declarative approach

Looks like this is actually picked up by JupyterLab too, which means that the shortcut is now disabled in JupyterLab too:

image

I wouldn't have expected that `jupyter.lab.shortcuts` configuration to be picked up if the plugin is not loaded on the page. Gonna check whether this is an issue in lab, or if we can find a workaround here in the meantime.

**Comment by jtpio** [comment]
> I wouldn't have expected that jupyter.lab.shortcuts configuration to be picked up if the plugin is not loaded on the page. Gonna check whether this is an issue in lab, or if we can find a workaround here in the meantime.

This is now tracked in https://github.com/jupyterlab/jupyterlab/issues/11754.

Based on these observations, the least surprising for the end user would then be to keep the programmatic shortcut override added in https://github.com/jupyterlab/retrolab/pull/294/commits/38a6166b23d31ac252b4c3f8209a21b6198ee66e. So it doesn't have any side-effect in JupyterLab.

I'll update this PR to revert the last two commits, so we can get it out in the next release and revisit later depending on the status of https://github.com/jupyterlab/jupyterlab/issues/11754.

**Comment by jtpio** [comment]
Folks can use the Advanced Settings editor to re-enable the shortcut if they would like to (opt-in):

https://user-images.githubusercontent.com/591645/147565179-514ee94e-6802-4ebd-a971-b3013dd66bbd.mp4

**Review by jtpio — APPROVED** [review]
Thanks!

**Comment by loichuder** [comment]
Thanks for following that up and ironing out the last wrinkles :slightly_smiling_face:  :+1:

**Comment by jtpio** [comment]
@loichuder thank **you** for working on this!

FYI this change was released in `0.3.14`: https://github.com/jupyterlab/retrolab/releases/tag/v0.3.14

**Pull request merged**